[ready for review] [scan] Refactor matching simulator devices for `device` flags and defaults.

jdhealy opened the pull request:
- Replace untrue comment on `filter_simulators` to reflect function behavior.
- Match only simulators with versions ‘that are greater than or equal to the version of our deployment target.’
- Be less naïve with striping parentheses. Only match versions at end of line.
- Unify logic for iOS and tvOS.

Fixes #5742.

jdhealy committed:
[scan] Refactor matching simulator devices for `device` flags and defaults.

- For arguments of `device`(s) flag with a version specifier, filter through only simulators with same version.

- For arguments of `device`(s) flag without a version specifier, (for example `--device 'iPad 2'`), return the single highest versioned simulator device available with matching name — [based off behavior pre-`devices` flag][1], which would result in a single simulator, but unfiltered on version.

- Replace untrue comment on `filter_simulators` to reflect function behavior.
- Match only simulators with versions ‘that are greater than or equal to the version of our deployment target.’
- Be less naïve with striping parentheses. Only match versions at end of line.
- Unify logic for iOS and tvOS.

- Improve error reporting.

[1]: https://github.com/fastlane/fastlane/blob/c23f35790ac49d2b96978f5200d71eae85995501/scan/lib/scan/detect_values.rb

jdhealy committed:
[scan] Refactor matching simulator devices for `device` flags and defaults.

- Replace untrue comment on `filter_simulators` to reflect function behavior.
- Match only simulators with versions ‘that are greater than or equal to the version of our deployment target.’
- Be less naïve with striping parentheses. Only match versions at end of line.
- Unify logic for iOS and tvOS.

jdhealy committed:
Un-nest method definition.

Addresses [hound-ci comment](https://github.com/fastlane/fastlane/pull/5746/files#r74694528).

[ci skip]

jdhealy committed:
Style: Prefer `inject` over `reduce`.

Addresses [hound-ci comment](https://github.com/fastlane/fastlane/pull/5746/files#r74694523).

[ci skip]

jdhealy committed:
Style: Avoid using `{...}` for multi-line blocks.

Addresses [hound-ci comment](https://github.com/fastlane/fastlane/pull/5746/files#r74694524).

[ci skip]

jdhealy committed:
Style: Do not use parentheses for method calls with no arguments.

Addresses [hound-ci comment](https://github.com/fastlane/fastlane/pull/5746/files#r74694521).

[ci skip]

jdhealy committed:
Correct wrong number of arguments.

<https://circleci.com/gh/fastlane/fastlane/3796>

jdhealy committed:
Return expression rather than addition assignment.

jdhealy committed:
Remove accidentally leftover `default_device_tvos`.

jdhealy committed:
Replace `flat_map` with desired `reject` of `nil` values.

jdhealy committed:
Lift devices config array to function scope.

It's needed in another lambda.

jdhealy committed:
Preface `.first` call with `count` check.

This shouldn’t make any functional difference, but is visually a little more type-safe.

jdhealy committed:
Replace multi-line commented regular expression.

jdhealy committed:
Style: Rubocop: Remove non-ASCII characters from comments.

jdhealy committed:
Style: Rubocop: Avoid using `{...}` with local variable and lambda.

ohwutup commented:
Thanks for the PR @jdhealy! This looks like quite a change to the way we handle simulators, would you mind doing two little things to get this to move forward? 

1. Could you fix up the hound/rubocop failures?
2. Would you mind adding a couple of tests to show where this was failing before and to validate the new behavior? 

Thank you!! 🐙 🚀

jdhealy committed:
Style: Rubocop: Use arrow symbol for lambda with arguments.

Also, remove trailing whitespace.

jdhealy committed:
Refactor `filter_simulator`.

- For arguments of `device`(s) flag with a version specifier, filter through only simulators with same version.

- For arguments of `device`(s) flag without a version specifier, (for example `--device 'iPad 2'`), return the single highest versioned simulator device available with matching name — [based off behavior pre-`devices` flag][1], which would result in a single simulator, but unfiltered on version.

- Improve error reporting.

[1]: https://github.com/fastlane/fastlane/blob/c23f35790ac49d2b96978f5200d71eae85995501/scan/lib/scan/detect_values.rb

jdhealy committed:
Default to single highest versioned iOS simulator device available matching default device name.

jdhealy committed:
[scan] Improve command generator specs.

Add more iOS versions.

Make sure it:
- “uses a device without version specifier”
- “rejects devices with versions below deployment target”
- “de-duplicates devices matching same simulator”

jdhealy committed:
[scan] Improve command generator specs.

Add more iOS versions.

Make sure it:
- “uses a device without version specifier”
- “rejects devices with versions below deployment target”
- “de-duplicates devices matching same simulator”

jdhealy committed:
Syntax: Style: un-hang conditional OR operator.

jdhealy committed:
Style: Rubocop: Strings and lambdas.

jdhealy commented:
Behavior should now be more in line with [pre-`devices`-flag behavior](https://github.com/fastlane/fastlane/blob/c23f35790ac49d2b96978f5200d71eae85995501/scan/lib/scan/detect_values.rb):

- For arguments of `device`(s) flag with a version specifier, filter through only simulators with same version.

- For arguments of `device`(s) flag without a version specifier, (for example `--device 'iPad 2'`), return the single highest versioned simulator device available with matching name.

- Default to single highest versioned iOS simulator device available matching default device name (`iPhone 5s` or `Apple TV 1080p` or whatever).

jdhealy commented on scan/spec/test_command_generator_spec.rb:
```diff
@@ -166,6 +198,32 @@
+      Scan.config = FastlaneCore::Configuration.create(Scan::Options.available_options, options)
+
+      result = Scan::TestCommandGenerator.generate
+      # FIXME: expect UI error starting "No simulators found that are equal to the version of specifier"
```
Not quite sure how to make `rspec` check for this. Could use some help with this…

ohwutup replied:
Happy to help! 

Are you expecting an exception to be raised here? What is it that you are looking to validate exactly?

jdhealy replied:
@ohwutup [this UI.error: detect_values.rb#L117-L119](https://github.com/jdhealy/fastlane/blob/03d8cb3bf0f0d871c595f6c1ed0df985100d31d2/scan/lib/scan/detect_values.rb#L117-L119)

ohwutup replied:
Ah! Makes sense... So I think the best way to handle that would be to call `UI.crash!(<message>)` instead of `UI.error` at the place of the link you shared, because if there is no simulator found, we can't proceed anyway right? 

That would also make this test easy to implement with something like this in `rspec`: https://www.relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/raise-error-matcher

jdhealy replied:
> because if there is no simulator found, we can't proceed anyway right?

Not for that location, at least. The error could show for one device in a `--devices` array, but the next `device_string` could resolve. In a different scenario, it could select the default.

Is there a `UI.warning` that would be a better fit than `.error`? Does error end the process?

jdhealy committed:
Revert previous 20 commits for upcoming squashed commit.

jdhealy commented:
With the exception of [the rspec question](https://github.com/fastlane/fastlane/pull/5746#discussion_r74979397), this is set and ready for review. 🌈

jdhealy commented:
@ohwutup Notes, comments, review?

I believe this addresses #5742 — which @postmechanical [confirmed to still be an issue in `scan-0.11.3`](https://github.com/fastlane/fastlane/issues/5731#issuecomment-241087446).

jdhealy commented:
welp, [tried to make headway](https://github.com/jdhealy/fastlane/blob/160994ea857f83c11e1f0fbe95cef9ea478ef4eb/scan/spec/test_command_generator_spec.rb#L219-L227) on [that one rspec question](https://github.com/fastlane/fastlane/pull/5746#discussion_r74979397), but [no dice](https://circleci.com/gh/fastlane/fastlane/4086).

So, back to the same diff as [jdhealy/fastlane@`03d8cb`](https://github.com/jdhealy/fastlane/blob/03d8cb3bf0f0d871c595f6c1ed0df985100d31d2/scan/lib/scan/detect_values.rb), but squashed down, so the last two commits can be easily cherry-picked for clean history.

Code works, just unversed in how to finagle `rspec` to expect that `UI.error`.

KrauseFx commented:
Hey @jdhealy, really sorry for the delay from our side, I'm looking into this issue and your PR right now 👍

KrauseFx commented:
@jdhealy can you drop me a DM on Twitter (@KrauseFx) with your email? 👍

KrauseFx commented:
This is great, thanks for working on this @jdhealy, I'll add a follow-up commit in a few minutes 👍
